V1 of bedrock-oauth2-client.

mandyvenables opened the pull request:
A client that takes in the Oauth2 data and returns an accessToken.

Addresses https://github.com/digitalbazaar/veres-wallet/issues/367

mandyvenables committed:
Create oauth2-client.js

mandyvenables committed:
Update package.json

mandyvenables committed:
Update year to 2021.

mandyvenables committed:
Update package.json

mandyvenables committed:
Update class name.

mandyvenables committed:
Fix variable names.

mandyvenables committed:
Remove config and http files.

mandyvenables committed:
Add expiration.

mandyvenables committed:
Update `checkDatabase`.

mandyvenables committed:
Update CHANGELOG.md

dmitrizagidulin commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+      agent: httpsAgent
+    }));
+  } catch(e) {
+    throw new BedrockError(
```
Can we add an error log entry here as well, in addition to throwing the BedrockError?
Returning an opaque 503 Service Unavailable to whoever is trying to use this endpoint is totally fine,
but we also want to make sure that our error logs contain the reason "This service is unavailable because we couldn't get a critical access_token - and here's the reason that operation failed."

mandyvenables replied:
Added logger.

dmitrizagidulin commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+  }
+}
+
+async function getNewAccessToken({
```
It'd be great to have jsdocs, here and on other functions.

mandyvenables replied:
Added!

dmitrizagidulin commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+    if(!database.isDuplicateError(e)) {
+      throw e;
+    }
+    throw new BedrockError(
```
Do we actually need to throw a 'Duplicate token' error here?

mandyvenables replied:
Hmm, I don't think we do need to throw a DuplicateError, removing.

dmitrizagidulin commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+    this.scope = scope;
+    this.httpsAgent = httpsAgent;
+  }
+  async getToken() {
```
```suggestion
  async loadToken() {
```
Would `loadToken` be more descriptive? (Emphasizing that we're loading it from db first)

dlongley replied:
I haven't looked more closely at this PR yet, but our naming convention for getting things from the database is `get` not `load`.

dmitrizagidulin commented on test/mocha/10-api.js:
```diff
@@ -1,5 +1,5 @@
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
  */
 
 describe('api', () => {
```
Can we add unit tests (for checkDatabase, getToken, insertToken etc)?

dmitrizagidulin commented:
This is a good start. A few more comments:

1. Let's move the parsing of config into this library (currently it's being done in the top-level app).
2. Add the ability to store different access tokens by issuer
3. Add the memoized LRU cache of access tokens
4. Add the max retry logic.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+
```
```suggestion
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+const {util: {BedrockError}} = bedrock;
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['access-tokens']);
```
We don't need `promisify` here anymore.

```suggestion
  await database.openCollections(['access-tokens']);
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+
+import database from 'bedrock-mongodb';
+import * as bedrock from 'bedrock';
+import {promisify} from 'util';
```
```suggestion
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+  await promisify(database.openCollections)(['access-tokens']);
+
+  // shard based on `siteId`
+  await promisify(database.createIndexes)([{
```
```suggestion
  await database.createIndexes([{
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,131 @@
+    fields: {'token.token': 1},
+    options: {unique: true, background: false}
+  }, {
+    // automatically expire events with an `expires` date field
```
These comments talk about `events` not `tokens`.

dlongley reviewed (commented):
Not a full review from me yet, just some partial comments thus far.

mandyvenables committed:
Remove duplicate error.

mandyvenables committed:
Add logger, remove unneeded else.

mandyvenables committed:
Add jsdocs.

mandyvenables committed:
Add basic tests.

mandyvenables committed:
Parse config in `OAuth2Client`.

mandyvenables commented:
@dmitrizagidulin  I have added in basic tests and moved parsing of the config to this library. I have a few questions about the other items:

1. Add the ability to store different access tokens by issuer:
 - Currently we are storing the `client_id` as the identifier in the database, what other properties would need to be stored to achieve this? `client_id` and `url`?

2. Add the memoized LRU cache of access tokens
 - A concern with the cache is not knowing if the accessToken has expired, since if it was being consistently fetched it could stay in the cache past expiration date. We can store the expiration date in the cache and check that it's not expired before sending it back; but would that extra step negate the benefit of using the cache? Also, in theory this will only be called on start up and every 2 weeks, is there a benefit to using the cache at all?

mandyvenables committed:
Add maxRetries logic.

dmitrizagidulin commented:
@mandyvenables - the access token cache is intended to be for managing clients from multiple issuers, but I think I'm mixing abstractions. So, this first iteration, let's just narrow it to a single issuer / client registration. Ignore my comments 1 & 2.

mandyvenables committed:
Change `url` to `token_endpoint`.

mandyvenables commented:
@dlongley @dmitrizagidulin  this is ready for an in depth review when you have time.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+    --maxRetries;
+  }
+  throw new BedrockError(
+    `Service Unavilable`, 'InvalidStateError', {
```
It strikes again!

```suggestion
    `Service Unavailable`, 'InvalidStateError', {
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+      agent: httpsAgent
+    }));
+  } catch(e) {
+    logger.error('Error getting accessToken.', {error: e});
```
```suggestion
    logger.error('Error getting access token.', {error: e});
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.token - The token record, including client id and
+ * accessToken.
```
```suggestion
 *   access token.
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+    });
+}
+
+async function httpCall({url, body, httpsAgent}) {
```
Function names should be verbs; `httpCall` is a noun vs. `httpPost` is a verb.

```suggestion
async function httpPost({url, body, httpsAgent}) {
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+ *
+ * @returns {string} The accessToken.
+ */
+async function getNewAccessToken({
```
Move all non-exported helper functions below all exported functions so that someone reading the file will hit the public exports first.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+ *
+ * @returns {string} The accessToken if there is one, null if not.
+ */
+export async function checkDatabase({client_id}) {
```
What does `checkDatabase` mean? This looks like it should be `hasClient` ... but then I see the token is returned as well, so it should be `getAccessToken`.

EDIT: And seeing that we already have `getToken` above, this function should be moved into `storage.js` and turned into simply `get()`.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+ * @param {object} options - Options to use.
+ * @param {string} options.client_id - The ID of the client.
+ *
+ * @returns {string} The accessToken if there is one, null if not.
```
Is `accessToken` camel case a term from a spec? Or should it be `access token` everywhere in the docs?

```suggestion
 * @returns {string} The access token if there is one, null if not.
```

mandyvenables replied:
Updated to `access token`.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+const {util: {BedrockError}} = bedrock;
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await database.openCollections(['access-tokens']);
```
This collection name doesn't follow our naming convention. It should be: `<bedrock module name prefix>-<camel case local collection name that reflects a singular doc name>` (eg: https://github.com/digitalbazaar/bedrock-tokenization/blob/master/lib/documents.js#L22). So this should be: `oauth2-client-accessToken`.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+   *
+   * @returns {string} The accessToken.
+   */
+  async getToken() {
```
Now that I see that we have a different method that actually gets a token from the database, this should be changed. What I'd recommend is moving all of the database functions into a separate file: `storage.js` and making them use the naming convention `insert()`, `get()`, etc. Then the main file can continue using the name that's being used here. That way it's clear when you're calling `storage.get()` that you're hitting storage to get something. That's the model we use in a number of other places.

dlongley replied:
This pattern would also make things easier to review and maintain as it has a clear separation of concerns via different files.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,185 @@
+    'token.client_id': client_id
+  };
+  let token;
+  const result = await database.collections['access-tokens'].findOne(query);
```
The naming convention we use for database records is `record`. If you make a mongodb call and it returns something other than the record itself, then `result` is a good name. But `findOne` actually just returns the `record` directly, I believe -- with no wrapper that has other result information.

```suggestion
  const record = await database.collections['access-tokens'].findOne(query);
```

mandyvenables committed:
Style updates.

mandyvenables committed:
Move database functions to `storage`.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import * as storage from './storage';
```
Always use file extensions when using ESM style imports to import files (as opposed to packages).

```suggestion
import * as storage from './storage.js';
```

mandyvenables replied:
Updated.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+import * as storage from './storage';
+import * as bedrock from 'bedrock';
+import {httpClient} from '@digitalbazaar/http-client';
+import logger from './logger';
```
```suggestion
import logger from './logger.js';
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+ * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+ *
+ * @returns {OAuth2Client} The `OAuth2Client` instance.
+ */
```
```suggestion
 /**
  * Constructs an OAuth2Client instance to return an access token. It first
  * checks the database to see if an access token already exists for the client
  * id; if not, it creates an access token then stores it in the database.
  *
  * @param {object} options - Options to use.
  * @param {string} options.config - The "oauth2-client-grant" properties.
  * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
  *
  * @returns {OAuth2Client} The `OAuth2Client` instance.
  */
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+    this.maxRetries = config.maxRetries;
+    this.httpsAgent = httpsAgent;
+  }
+  /**
```
```suggestion
  }

  /**
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+      maxRetries
+    } = this;
+    let token;
+    token = await storage.get({client_id});
```
Consolidating to a single line and using destructuring due to a suggested change below to the `storage.get` API.

```suggestion
    let {token} = await storage.get({client_id}) || {};
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+    } = this;
+    let token;
+    token = await storage.get({client_id});
+    if(!token || !token.access_token) {
```
```suggestion
    if(!(token && token.access_token)) {
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+   *
+   * @returns {string} The access token.
+   */
+  async getToken() {
```
There's an inconsistency in the API here where sometimes `Token` is used in the function name and other times `AccessToken` is. This should be normalized. I recommend using `AccessToken` since there are other types of tokens in oauth.

```suggestion
  async getAccessToken() {
```

mandyvenables replied:
Agreed, updated.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+  body.set('client_id', client_id);
+  body.set('client_secret', client_secret);
+  body.set('grant_type', grant_type);
+  body.set('scope', scope);
```
I think this should work:

```suggestion
  const body = new URLSearchParams(
    {client_id, client_secret, grant_type, scope});
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+      return token;
+    }
+    --maxRetries;
+  }
```
Declare vars closest to where they are needed and only where needed. Here `response` is should go down with the loop where it will be set, however, it actually isn't even necessary if we update `httpPost` to return the access token directly. Additionally, `token` isn't needed unless the conditional within the loop actually happens. Even then, we don't need the extra var, we can just return the full token at that time.

The code can be additionally shortened by using a for loop. It's also good to leave some white space after the loop to break up the loop section from the section that throws.

```suggestion
  for(; maxRetries >= 0; --maxRetries) {
    const access_token = await _createAccessToken(
      {url: token_endpoint, body, httpsAgent});
    if(access_token) {
      return {client_id, client_secret, access_token};
    }
  }

```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+    });
+}
+
+async function httpPost({url, body, httpsAgent}) {
```
Looking at the implementation of this helper, it is specific to creating an access token, it is not just an `httpPost`, so let's rename it. Also, it's an internal helper, so let's add an `_` prefix.

```suggestion
async function _createAccessToken({url, body, httpsAgent}) {
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,113 @@
+    logger.error('Error getting access token.', {error: e});
+  }
+  if(response && response.access_token) {
+    return response;
```
```suggestion
    return response.access_token;
```

dlongley commented on lib/storage.js:
```diff
@@ -0,0 +1,78 @@
+ * @returns {string} The access token if there is one, null if not.
+ */
+export async function get({client_id}) {
+  const query = {
+    'token.client_id': client_id
+  };
+  let token;
+  const record =
+    await database.collections['oauth2-client-accessToken'].findOne(query);
+  if(record && record.token) {
+    token = record.token;
+  }
+  return token;
+}
```
We don't need this extra `token` var. We can just return the token if it is found. Also, this function does not return `null` right now when there is no token (it does not match the docs). It returns `undefined`. This suggestion updates that too -- however, it also just returns the full record. The storage API should not assume what people want to do with the records and it's cleaner to give them the whole record and it mirrors `insert()` (note: this also follows our standard practice for storage APIs). We very frequently find that we have to refactor to provide the whole record in the future for various reasons anyway.

Also, don't break after equals, break on parenthesis or on `.` separator -- or use a var to hold the collection name as that often looks clean too.

```suggestion
  const collection = database.collections['oauth2-client-accessToken'];
  const query = {'token.client_id': client_id};
  return collection.findOne(query);
}
```

dlongley commented on lib/storage.js:
```diff
@@ -0,0 +1,78 @@
+export async function insert({token}) {
+  const now = Date.now();
+  const meta = {created: now, updated: now};
+  // expires after 2 weeks
+  const ttl = 2 * 7 * 24 * 60 * 60 * 1000;
+  token.expires = new Date(now + ttl);
+  let record = {
+    meta,
+    token
+  };
```
Better not to modify the value passed into the function:

```suggestion
  // token expires after 2 weeks
  const ttl = 2 * 7 * 24 * 60 * 60 * 1000;
  let record = {
    meta,
    token: {
      ...token,
      expires: new Date(now + ttl)
    }
  };
```

mandyvenables committed:
Style updates.

dlongley commented on lib/main.js:
```diff
@@ -1,7 +1,6 @@
-import './config.js';
-import './http.js';
+export {OAuth2Client} from './oauth2-client.js';
+export {get, insert} from './storage.js';
```
So looking more closely at this, I think it's odd to export `get` and `insert` directly like this. We should either export the full storage API using the name `storage` or not export it at all.

mandyvenables replied:
Updated.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,107 @@
+
+  /**
+   * Gets a new access token by first checking the database to see if an
+   *  access token already exists for the client id; if not it creates an
+   *  access token then stores it in the database.
```
```suggestion
   * access token already exists for the client id; if not it creates an
   * access token then stores it in the database.
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,107 @@
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries
+    } = this;
+    let {token} = await storage.get({client_id}) || {};
+    if(!(token && token.access_token)) {
+      token = await getNewAccessToken({
+        client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+        maxRetries
+      });
+      await storage.insert({token});
+    }
```
So this needs to be wrapped using lru-memoize so that when a token isn't available in the database, multiple calls aren't made to get a new one. We've used that pattern many times, one example to look at is here:

https://github.com/digitalbazaar/bedrock-kms/blob/main/lib/keystores.js#L188-L189

I believe we'd want to create a helper function here that does all the things I highlighted and that has some configurable max age set on it with a sane default.

mandyvenables replied:
Added! What is a "sane default"? I started with 10, but I'm not sure what the expected ballpark is.

Also, what should the maxAge be? I set it to 2 weeks, but I think that might not be correct since there could be a few second overlap between the token expiring but still being in the cache.

dmitrizagidulin replied:
@mandyvenables I think 2 weeks is fine for now.

dlongley reviewed (changes requested):
Getting very close! Left a couple other comments.

mandyvenables committed:
Rename exports.

mandyvenables committed:
Add cache.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,123 @@
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries, cacheMax = 10
+    } = this;
+    const cache = new LruCache({
+      max: cacheMax,
+      // 2 weeks
+      maxAge: 2 * 7 * 24 * 60 * 60 * 1000
+    });
```
So this cache needs to be created at the module level (you can take a look at the pattern I linked to before for how to  name it and set it up w/a bedrock event handler, etc). Doing it here will just cause it to go out of scope and get garbage collected as soon as `getAccessToken()` returns.

Also, the config should come from the `bedrock.config` for this module. As for the defaults, we want to be able to store as many access tokens as a single service would be reasonably using -- without more knowledge about that I'd just go with `100`. As for the max age, this cache is only used to avoid hitting the authz server and the database too frequently -- it isn't related to how long the access tokens live. I'd set it to 5 minutes by default.

mandyvenables replied:
Updated, thanks for the clarification.

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,123 @@
+      // 2 weeks
+      maxAge: 2 * 7 * 24 * 60 * 60 * 1000
+    });
+    const fn = () => retrieveAccessToken({
```
I don't think we need to expose this method for public use, so let's prefix it with `_` to mark it as something people should not call/rely on. Also, the name of the helper seems to just be redundant with `getAccessToken`. Let's name it after the fact that it is supposed to avoid the cache:

```suggestion
    const fn = () => _getUncachedAccessToken({
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,123 @@
+  }
+}
+
+async function retrieveAccessToken({
```
```suggestion
async function _getUncachedAccessToken({
```

mandyvenables committed:
Move cache and config.

mandyvenables committed:
Remove `cacheMax`.

dlongley commented on lib/config.js:
```diff
@@ -1,13 +1,13 @@
-cfg.routes = {
-  basePath
+config['oauth2-client'] = {
+  cache: {
```
We may need some other caching in the future, so let's make this more specific:

```suggestion
  accessTokenCache: {
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,127 @@
+import logger from './logger.js';
+const {util: {BedrockError}} = bedrock;
+
+let cache;
```
```suggestion
let ACCESS_TOKEN_CACHE;
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,127 @@
+
+bedrock.events.on('bedrock.init', async () => {
+  const cfg = bedrock.config['oauth2-client'];
+  cache = new LruCache({
+    max: cfg.cache.maxSize,
+    maxAge: cfg.cache.maxAge
+  });
```
```suggestion
  ACCESS_TOKEN_CACHE = new LruCache({
    max: cfg.accessTokenCache.maxSize,
    maxAge: cfg.accessTokenCache.maxAge
  });
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,127 @@
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries
+    } = this;
+    const fn = () => _getUncachedAccessToken({
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries});
```
Style fix: if the opening brace `{` is left on the same line, then drop the closing brace `}` to its own line as well:

```suggestion
    const fn = () => _getUncachedAccessToken({
      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
      maxRetries
    });
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,127 @@
+   * access token already exists for the client id; if not it creates an
+   * access token then stores it in the database.
+   *
+   * @returns {string} The access token.
```
Should this include `Promise` (here and in any other async functions)?

```suggestion
   * @returns {Promise<string>} The access token.
```

dlongley commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,127 @@
+    const fn = () => _getUncachedAccessToken({
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries});
+    return cache.memoize({key: client_id, fn});
```
```suggestion
    return ACCESS_TOKEN_CACHE.memoize({key: client_id, fn});
```

dlongley commented on lib/storage.js:
```diff
@@ -0,0 +1,73 @@
+  const now = Date.now();
+  const meta = {created: now, updated: now};
+  // token expires after 2 weeks
+  const ttl = 2 * 7 * 24 * 60 * 60 * 1000;
```
This `ttl` could also be part of the config for the lib so it can be changed if needed.

dlongley reviewed (approved):
Just a couple more minor comments to address but approving presuming those will be easy. Thanks!

dmitrizagidulin commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,127 @@
+  });
+
+  for(; maxRetries >= 0; --maxRetries) {
+    const access_token = await _createAccessToken(
```
```suggestion
    const access_token = await _requestAccessToken(
```

dmitrizagidulin commented on lib/oauth2-client.js:
```diff
@@ -0,0 +1,127 @@
+    });
+}
+
+async function _createAccessToken({url, body, httpsAgent}) {
```
```suggestion
async function _requestAccessToken({url, body, httpsAgent}) {
```

dmitrizagidulin reviewed (approved):
:+1:

mandyvenables committed:
Move ttl to config.

mandyvenables committed:
Update `bedrock-oauth2-client` name.

mandyvenables committed:
Update README.md
